{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## A. CellphoneDB pre-processing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "WARNING: If you miss a compact list, please try `print_header`!\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-----\n",
      "anndata     0.7.5\n",
      "scanpy      1.7.0rc1\n",
      "sinfo       0.3.1\n",
      "-----\n",
      "PIL                 8.1.2\n",
      "anndata             0.7.5\n",
      "anyio               NA\n",
      "attr                20.3.0\n",
      "babel               2.9.0\n",
      "backcall            0.2.0\n",
      "brotli              NA\n",
      "cairo               1.20.0\n",
      "certifi             2020.12.05\n",
      "cffi                1.14.5\n",
      "chardet             4.0.0\n",
      "cloudpickle         1.6.0\n",
      "colorama            0.4.4\n",
      "cycler              0.10.0\n",
      "cython_runtime      NA\n",
      "cytoolz             0.11.0\n",
      "dask                2021.03.0\n",
      "dateutil            2.8.1\n",
      "decorator           4.4.2\n",
      "get_version         2.1\n",
      "google              NA\n",
      "h5py                2.10.0\n",
      "idna                2.10\n",
      "igraph              0.9.1\n",
      "ipykernel           5.5.0\n",
      "ipython_genutils    0.2.0\n",
      "jedi                0.18.0\n",
      "jinja2              2.11.3\n",
      "joblib              1.0.1\n",
      "json5               NA\n",
      "jsonschema          3.2.0\n",
      "jupyter_server      1.4.1\n",
      "jupyterlab_server   2.3.0\n",
      "kiwisolver          1.3.1\n",
      "legacy_api_wrap     0.0.0\n",
      "leidenalg           0.8.3\n",
      "llvmlite            0.34.0\n",
      "louvain             0.7.0\n",
      "markupsafe          1.1.1\n",
      "matplotlib          3.3.4\n",
      "mpl_toolkits        NA\n",
      "natsort             7.1.1\n",
      "nbclassic           NA\n",
      "nbformat            5.1.2\n",
      "numba               0.51.2\n",
      "numexpr             2.7.3\n",
      "numpy               1.20.1\n",
      "packaging           20.9\n",
      "pandas              1.1.4\n",
      "parso               0.8.1\n",
      "pexpect             4.8.0\n",
      "pickleshare         0.7.5\n",
      "pkg_resources       NA\n",
      "prometheus_client   NA\n",
      "prompt_toolkit      3.0.16\n",
      "psutil              5.8.0\n",
      "ptyprocess          0.7.0\n",
      "pvectorc            NA\n",
      "pygments            2.8.0\n",
      "pyparsing           2.4.7\n",
      "pyrsistent          NA\n",
      "pytoml              NA\n",
      "pytz                2021.1\n",
      "requests            2.25.1\n",
      "ruamel              NA\n",
      "scanpy              1.7.0rc1\n",
      "scipy               1.6.1\n",
      "send2trash          NA\n",
      "setuptools_scm      NA\n",
      "sinfo               0.3.1\n",
      "six                 1.15.0\n",
      "sklearn             0.22\n",
      "sniffio             1.2.0\n",
      "socks               1.7.1\n",
      "sphinxcontrib       NA\n",
      "storemagic          NA\n",
      "tables              3.6.1\n",
      "tblib               1.7.0\n",
      "terminado           0.9.2\n",
      "texttable           1.6.3\n",
      "tlz                 0.11.0\n",
      "toolz               0.11.1\n",
      "tornado             6.1\n",
      "traitlets           5.0.5\n",
      "typing_extensions   NA\n",
      "urllib3             1.26.3\n",
      "wcwidth             0.2.5\n",
      "yaml                5.1.2\n",
      "zmq                 22.0.3\n",
      "-----\n",
      "IPython             7.21.0\n",
      "jupyter_client      6.1.11\n",
      "jupyter_core        4.7.1\n",
      "jupyterlab          3.0.9\n",
      "notebook            6.2.0\n",
      "-----\n",
      "Python 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 16:22:27) [GCC 9.3.0]\n",
      "Linux-4.15.0-112-generic-x86_64-with-glibc2.10\n",
      "26 logical CPU cores, x86_64\n",
      "-----\n",
      "Session information updated at 2021-04-24 14:32\n",
      "\n"
     ]
    }
   ],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import anndata\n",
    "import os\n",
    "import sys\n",
    "\n",
    "def MovePlots(plotpattern, subplotdir):\n",
    "    os.system('mkdir -p '+str(sc.settings.figdir)+'/'+subplotdir)\n",
    "    os.system('mv '+str(sc.settings.figdir)+'/*'+plotpattern+'** '+str(sc.settings.figdir)+'/'+subplotdir)\n",
    "\n",
    "sc.settings.verbosity = 1  # verbosity: errors (0), warnings (1), info (2), hints (3)\n",
    "sc.settings.figdir = './cellphoneDB/'\n",
    "sc.logging.print_versions()\n",
    "sc.settings.set_figure_params(dpi=80)  # low dpi (dots per inch) yields small inline figures\n",
    "\n",
    "sys.executable\n",
    "\n",
    "\n",
    "def grouped_obs_percent(adata, group_key, layer=None, gene_symbols=None):\n",
    "    if layer is not None:\n",
    "        getX = lambda x: x.layers[layer]\n",
    "    else:\n",
    "        getX = lambda x: x.X\n",
    "    if gene_symbols is not None:\n",
    "        new_idx = adata.var[idx]\n",
    "    else:\n",
    "        new_idx = adata.var_names\n",
    "\n",
    "    grouped = adata.obs.groupby(group_key)\n",
    "    out = pd.DataFrame(\n",
    "        np.zeros((adata.shape[1], len(grouped)), dtype=np.float64),\n",
    "        columns=list(grouped.groups.keys()),\n",
    "        index=adata.var_names\n",
    "    )\n",
    "\n",
    "    for group, idx in grouped.indices.items():\n",
    "        X = getX(adata[idx])\n",
    "        X.data = X.data > 0.01\n",
    "        perc = np.asarray(np.sum(X,axis=0)/X.shape[0]).reshape(-1)\n",
    "        out[group] = [round(i, 2) for i in perc ]\n",
    "    return out\n",
    "\n",
    "\n",
    "def grouped_obs_mean(adata, group_key, layer=None, gene_symbols=None):\n",
    "    if layer is not None:\n",
    "        getX = lambda x: x.layers[layer]\n",
    "    else:\n",
    "        getX = lambda x: x.X\n",
    "    if gene_symbols is not None:\n",
    "        new_idx = adata.var[idx]\n",
    "    else:\n",
    "        new_idx = adata.var_names\n",
    "\n",
    "    grouped = adata.obs.groupby(group_key)\n",
    "    out = pd.DataFrame(\n",
    "        np.zeros((adata.shape[1], len(grouped)), dtype=np.float64),\n",
    "        columns=list(grouped.groups.keys()),\n",
    "        index=adata.var_names\n",
    "    )\n",
    "\n",
    "    for group, idx in grouped.indices.items():\n",
    "        X = getX(adata[idx])\n",
    "        out[group] = np.ravel(X.mean(axis=0, dtype=np.float64))\n",
    "    return out"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Prepare INPUT\n",
    "\n",
    "## Load andata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(319081, 28230)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata = sc.read('/nfs/team292/lg18/with_valentina/FCA-M5-annotatedCluster4Seurat.h5ad')\n",
    "adata.X.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load cell clusters annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3165: DtypeWarning: Columns (24,25,27) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The lists are identical\n"
     ]
    }
   ],
   "source": [
    "# Add cluster name and doublet information\n",
    "clu_annot = pd.read_csv('figures_manual_annotation/clustering_metadata.csv', header=0, index_col=0)\n",
    "import collections \n",
    "\n",
    "if collections.Counter(adata.obs.index) == collections.Counter(clu_annot.index): \n",
    "    print (\"The lists are identical\") \n",
    "else : \n",
    "    print (\"The lists are not identical\") \n",
    "    \n",
    "vars2import = ['clusters_manual','louvain', 'is_doublet', 'scrublet_cluster_score', 'scrublet_score']\n",
    "for var in vars2import:\n",
    "    adata.obs[var] = clu_annot[var]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>5v1.1</th>\n",
       "      <th>TP</th>\n",
       "      <th>batch</th>\n",
       "      <th>batch.collection</th>\n",
       "      <th>cryopreserved</th>\n",
       "      <th>individual</th>\n",
       "      <th>location</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeatures_RNA</th>\n",
       "      <th>percent.mito</th>\n",
       "      <th>...</th>\n",
       "      <th>clusters_previous</th>\n",
       "      <th>S.Score</th>\n",
       "      <th>G2M.Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>old.ident</th>\n",
       "      <th>RNA_snn_res.0.3</th>\n",
       "      <th>seurat_clusters</th>\n",
       "      <th>clusters_old</th>\n",
       "      <th>clusters_sex</th>\n",
       "      <th>annotated_clusters</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>FCA_GND8047885_AAGACCTCAGTATAAG</th>\n",
       "      <td>2</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>F81</td>\n",
       "      <td>0</td>\n",
       "      <td>1469</td>\n",
       "      <td>3446</td>\n",
       "      <td>0.013602</td>\n",
       "      <td>...</td>\n",
       "      <td>PGC_mitotic</td>\n",
       "      <td>-0.061951</td>\n",
       "      <td>0.658402</td>\n",
       "      <td>G2M</td>\n",
       "      <td>FCA-M5-annotatedCluster4Seurat</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>PGC_mitotic</td>\n",
       "      <td>new</td>\n",
       "      <td>PGC_mitotic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCA_GND8047885_AAGGTTCAGTTAAGTG</th>\n",
       "      <td>2</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>F81</td>\n",
       "      <td>0</td>\n",
       "      <td>1763</td>\n",
       "      <td>3821</td>\n",
       "      <td>0.017654</td>\n",
       "      <td>...</td>\n",
       "      <td>PGC_mitotic</td>\n",
       "      <td>-0.126121</td>\n",
       "      <td>0.615875</td>\n",
       "      <td>G2M</td>\n",
       "      <td>FCA-M5-annotatedCluster4Seurat</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>PGC_mitotic</td>\n",
       "      <td>new</td>\n",
       "      <td>PGC_mitotic</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCA_GND8047885_ATCATCTTCGTTGACA</th>\n",
       "      <td>2</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>F81</td>\n",
       "      <td>0</td>\n",
       "      <td>850</td>\n",
       "      <td>2979</td>\n",
       "      <td>0.065334</td>\n",
       "      <td>...</td>\n",
       "      <td>PGC</td>\n",
       "      <td>-0.051857</td>\n",
       "      <td>-0.098312</td>\n",
       "      <td>G1</td>\n",
       "      <td>FCA-M5-annotatedCluster4Seurat</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>PGC</td>\n",
       "      <td>new</td>\n",
       "      <td>PGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCA_GND8047885_CCTCTGAAGTGACATA</th>\n",
       "      <td>2</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>F81</td>\n",
       "      <td>0</td>\n",
       "      <td>1446</td>\n",
       "      <td>4123</td>\n",
       "      <td>0.014408</td>\n",
       "      <td>...</td>\n",
       "      <td>PGC</td>\n",
       "      <td>-0.145298</td>\n",
       "      <td>-0.111084</td>\n",
       "      <td>G1</td>\n",
       "      <td>FCA-M5-annotatedCluster4Seurat</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>PGC</td>\n",
       "      <td>new</td>\n",
       "      <td>PGC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FCA_GND8047885_CGAACATAGCCGGTAA</th>\n",
       "      <td>2</td>\n",
       "      <td>U</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>F81</td>\n",
       "      <td>0</td>\n",
       "      <td>2458</td>\n",
       "      <td>4494</td>\n",
       "      <td>0.013342</td>\n",
       "      <td>...</td>\n",
       "      <td>PGC</td>\n",
       "      <td>-0.233330</td>\n",
       "      <td>-0.206078</td>\n",
       "      <td>G1</td>\n",
       "      <td>FCA-M5-annotatedCluster4Seurat</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>PGC</td>\n",
       "      <td>new</td>\n",
       "      <td>PGC</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 35 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 5v1.1 TP  batch  batch.collection  \\\n",
       "Unnamed: 0                                                           \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG      2  U      0                 0   \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG      2  U      0                 0   \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA      2  U      0                 0   \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA      2  U      0                 0   \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA      2  U      0                 0   \n",
       "\n",
       "                                 cryopreserved individual  location  \\\n",
       "Unnamed: 0                                                            \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG              0        F81         0   \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG              0        F81         0   \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA              0        F81         0   \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA              0        F81         0   \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA              0        F81         0   \n",
       "\n",
       "                                 nCount_RNA  nFeatures_RNA  percent.mito  ...  \\\n",
       "Unnamed: 0                                                                ...   \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG        1469           3446      0.013602  ...   \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG        1763           3821      0.017654  ...   \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA         850           2979      0.065334  ...   \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA        1446           4123      0.014408  ...   \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA        2458           4494      0.013342  ...   \n",
       "\n",
       "                                clusters_previous   S.Score G2M.Score  Phase  \\\n",
       "Unnamed: 0                                                                     \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG       PGC_mitotic -0.061951  0.658402    G2M   \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG       PGC_mitotic -0.126121  0.615875    G2M   \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA               PGC -0.051857 -0.098312     G1   \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA               PGC -0.145298 -0.111084     G1   \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA               PGC -0.233330 -0.206078     G1   \n",
       "\n",
       "                                                      old.ident  \\\n",
       "Unnamed: 0                                                        \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG  FCA-M5-annotatedCluster4Seurat   \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG  FCA-M5-annotatedCluster4Seurat   \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA  FCA-M5-annotatedCluster4Seurat   \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA  FCA-M5-annotatedCluster4Seurat   \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA  FCA-M5-annotatedCluster4Seurat   \n",
       "\n",
       "                                 RNA_snn_res.0.3 seurat_clusters  \\\n",
       "Unnamed: 0                                                         \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG                4               4   \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG                4               4   \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA                2               2   \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA                2               2   \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA                2               2   \n",
       "\n",
       "                                 clusters_old  clusters_sex  \\\n",
       "Unnamed: 0                                                    \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG   PGC_mitotic           new   \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG   PGC_mitotic           new   \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA           PGC           new   \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA           PGC           new   \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA           PGC           new   \n",
       "\n",
       "                                 annotated_clusters  \n",
       "Unnamed: 0                                           \n",
       "FCA_GND8047885_AAGACCTCAGTATAAG         PGC_mitotic  \n",
       "FCA_GND8047885_AAGGTTCAGTTAAGTG         PGC_mitotic  \n",
       "FCA_GND8047885_ATCATCTTCGTTGACA                 PGC  \n",
       "FCA_GND8047885_CCTCTGAAGTGACATA                 PGC  \n",
       "FCA_GND8047885_CGAACATAGCCGGTAA                 PGC  \n",
       "\n",
       "[5 rows x 35 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_som = pd.read_csv('/nfs/team292/lg18/with_valentina/supporting_nocycling_annotation.csv')\n",
    "meta_som = meta_som.set_index('Unnamed: 0')\n",
    "meta_som.head()\n",
    "\n",
    "meta_germ = pd.read_csv('/nfs/team292/lg18/with_valentina/germcells_annotation.csv')\n",
    "meta_germ = meta_germ.set_index('Unnamed: 0')\n",
    "meta_germ.head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Filter clusters of interest"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use only barcodes in metadata file containing germ or somatic cells in the gonad"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "barcodes = meta_som.index.tolist() + meta_germ.index.tolist() \n",
    "adata = adata[adata.obs.index.isin(barcodes),:]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "And add `cluster` label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "som_cell_annot = meta_som['annotated_clusters'].to_dict()\n",
    "germ_cell_annot = meta_germ['annotated_clusters'].to_dict()\n",
    "cell_annot = {**som_cell_annot, **germ_cell_annot}\n",
    "adata.obs['clusters'] = [ cell_annot[i] for i in adata.obs_names ] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FetalLeydig-like',\n",
       " 'GC_mitotic',\n",
       " 'PGC',\n",
       " 'PGC_mitotic',\n",
       " 'Sertoli',\n",
       " 'Sertoli_WFDC2',\n",
       " 'Sertoli_low',\n",
       " 'coelEpi',\n",
       " 'oocyte',\n",
       " 'oogonia_STRA8',\n",
       " 'oogonia_meiosis',\n",
       " 'ovarianSurf',\n",
       " 'pre-spermatogonia',\n",
       " 'preGC_II',\n",
       " 'preGC_III',\n",
       " 'preGC_III_Notch',\n",
       " 'preGC_II_hypoxia',\n",
       " 'preGC_I_OSR1',\n",
       " 'sKITLG',\n",
       " 'sLGR5',\n",
       " 'sPAX8b',\n",
       " 'sPAX8m'}"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata.obs['clusters'] )"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Add sex info in the bipotent clusters - to match microenviroments file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'FetalLeydig-like',\n",
       " 'GC_mitotic',\n",
       " 'PGC',\n",
       " 'PGC_mitotic',\n",
       " 'Sertoli',\n",
       " 'Sertoli_WFDC2',\n",
       " 'Sertoli_low',\n",
       " 'coelEpi',\n",
       " 'oocyte',\n",
       " 'oogonia_STRA8',\n",
       " 'oogonia_meiosis',\n",
       " 'ovarianSurf',\n",
       " 'pre-spermatogonia',\n",
       " 'preGC_II',\n",
       " 'preGC_III',\n",
       " 'preGC_III_Notch',\n",
       " 'preGC_II_hypoxia',\n",
       " 'preGC_I_OSR1',\n",
       " 'sKITLG',\n",
       " 'sLGR5_female',\n",
       " 'sLGR5_male',\n",
       " 'sPAX8b',\n",
       " 'sPAX8m'}"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = adata.obs['clusters'].isin(['sLGR5'])\n",
    "idx = s[s].index.values\n",
    "adata.obs['clusters_sex'] =  adata.obs['clusters']\n",
    "for i in idx:\n",
    "    adata.obs.at[i,'clusters_sex'] = adata.obs['clusters'][i]+'_'+adata.obs['sex'][i]\n",
    "set(adata.obs['clusters_sex'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Filter cells in the microenviroments of interest"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>celltype</th>\n",
       "      <th>microenviroment</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>coelEpi</td>\n",
       "      <td>cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ovarianSurf</td>\n",
       "      <td>cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>PGC</td>\n",
       "      <td>cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>PGC_mitotic</td>\n",
       "      <td>cortex</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>preGC_II_hypoxia</td>\n",
       "      <td>cortex</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           celltype microenviroment\n",
       "0           coelEpi          cortex\n",
       "1       ovarianSurf          cortex\n",
       "2               PGC          cortex\n",
       "3       PGC_mitotic          cortex\n",
       "4  preGC_II_hypoxia          cortex"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "micro = pd.read_csv('cellphoneDB/input/microenviroments/microenviroments.csv')\n",
    "# micro = micro.set_index('Unnamed: 0')\n",
    "micro.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "cells_common = list(set(adata.obs['clusters_sex'] )  & set(micro.celltype)) \n",
    "adata = adata[[ i in cells_common for i in adata.obs.clusters_sex ]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'PGC',\n",
       " 'PGC_mitotic',\n",
       " 'Sertoli',\n",
       " 'Sertoli_WFDC2',\n",
       " 'coelEpi',\n",
       " 'oocyte',\n",
       " 'oogonia_STRA8',\n",
       " 'oogonia_meiosis',\n",
       " 'ovarianSurf',\n",
       " 'pre-spermatogonia',\n",
       " 'preGC_II',\n",
       " 'preGC_III',\n",
       " 'preGC_III_Notch',\n",
       " 'preGC_II_hypoxia',\n",
       " 'preGC_I_OSR1',\n",
       " 'sKITLG',\n",
       " 'sLGR5_female',\n",
       " 'sLGR5_male',\n",
       " 'sPAX8b',\n",
       " 'sPAX8m'}"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(adata.obs['clusters_sex'] )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "# pd.DataFrame(adata.obs['clusters_sex']).to_csv(str(sc.settings.figdir)+'/input/metadata.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Compute average and percentatge expression per cluster"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalized average\n",
    "adata_norm = adata.copy()\n",
    "sc.pp.normalize_per_cell(adata_norm, counts_per_cell_after=1e4)\n",
    "sc.pp.log1p(adata_norm)\n",
    "means = grouped_obs_mean(adata_norm, 'clusters_sex')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "# percentatge expression\n",
    "percent = grouped_obs_percent(adata, 'clusters_sex')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# # un-byte\n",
    "# means.index = [ i.decode(\"utf-8\") for i in means.index]\n",
    "# percent.index = [ i.decode(\"utf-8\") for i in percent.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PGC</th>\n",
       "      <th>PGC_mitotic</th>\n",
       "      <th>Sertoli</th>\n",
       "      <th>Sertoli_WFDC2</th>\n",
       "      <th>coelEpi</th>\n",
       "      <th>oocyte</th>\n",
       "      <th>oogonia_STRA8</th>\n",
       "      <th>oogonia_meiosis</th>\n",
       "      <th>ovarianSurf</th>\n",
       "      <th>pre-spermatogonia</th>\n",
       "      <th>preGC_II</th>\n",
       "      <th>preGC_III</th>\n",
       "      <th>preGC_III_Notch</th>\n",
       "      <th>preGC_II_hypoxia</th>\n",
       "      <th>preGC_I_OSR1</th>\n",
       "      <th>sKITLG</th>\n",
       "      <th>sLGR5_female</th>\n",
       "      <th>sLGR5_male</th>\n",
       "      <th>sPAX8b</th>\n",
       "      <th>sPAX8m</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>MIR1302-2HG</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL669831.5</th>\n",
       "      <td>0.11</td>\n",
       "      <td>0.14</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.07</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.06</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.05</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FAM41C</th>\n",
       "      <td>0.02</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.01</td>\n",
       "      <td>0.08</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.02</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              PGC  PGC_mitotic  Sertoli  Sertoli_WFDC2  coelEpi  oocyte  \\\n",
       "index                                                                     \n",
       "MIR1302-2HG  0.00         0.00     0.00           0.00     0.00    0.00   \n",
       "AL627309.1   0.00         0.00     0.00           0.00     0.00    0.07   \n",
       "AL669831.5   0.11         0.14     0.01           0.01     0.05    0.07   \n",
       "LINC00115    0.03         0.04     0.03           0.05     0.04    0.02   \n",
       "FAM41C       0.02         0.03     0.00           0.00     0.01    0.08   \n",
       "\n",
       "             oogonia_STRA8  oogonia_meiosis  ovarianSurf  pre-spermatogonia  \\\n",
       "index                                                                         \n",
       "MIR1302-2HG           0.00             0.00         0.00               0.01   \n",
       "AL627309.1            0.01             0.02         0.00               0.01   \n",
       "AL669831.5            0.09             0.06         0.06               0.03   \n",
       "LINC00115             0.06             0.04         0.03               0.04   \n",
       "FAM41C                0.04             0.04         0.00               0.02   \n",
       "\n",
       "             preGC_II  preGC_III  preGC_III_Notch  preGC_II_hypoxia  \\\n",
       "index                                                                 \n",
       "MIR1302-2HG      0.00       0.00             0.00              0.00   \n",
       "AL627309.1       0.00       0.00             0.00              0.00   \n",
       "AL669831.5       0.04       0.03             0.04              0.07   \n",
       "LINC00115        0.03       0.05             0.05              0.03   \n",
       "FAM41C           0.00       0.00             0.00              0.00   \n",
       "\n",
       "             preGC_I_OSR1  sKITLG  sLGR5_female  sLGR5_male  sPAX8b  sPAX8m  \n",
       "index                                                                        \n",
       "MIR1302-2HG          0.00    0.00          0.00        0.00    0.00    0.00  \n",
       "AL627309.1           0.00    0.00          0.00        0.00    0.00    0.00  \n",
       "AL669831.5           0.04    0.04          0.05        0.03    0.04    0.02  \n",
       "LINC00115            0.04    0.03          0.05        0.04    0.04    0.04  \n",
       "FAM41C               0.00    0.00          0.00        0.00    0.00    0.00  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "means.head()\n",
    "percent.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "pd.DataFrame(means).to_csv(str(sc.settings.figdir)+'/input/average_log.csv')\n",
    "pd.DataFrame(percent).to_csv(str(sc.settings.figdir)+'/input/percent.csv')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save gene expression \n",
    "\n",
    "Raw counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/lib/python3.8/site-packages/anndata/_core/anndata.py:1207: ImplicitModificationWarning: Initializing view as actual.\n",
      "  warnings.warn(\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'clusters' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'clusters_manual' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'clusters_sex' as categorical\n"
     ]
    }
   ],
   "source": [
    "adata.write(str(sc.settings.figdir)+'/input/counts.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Save meta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_meta = pd.DataFrame(data={'Cell':list(adata.obs.index),\n",
    "                             'cell_type':[ 'celltype_'+str(i) for i in adata.obs['clusters_sex']] })\n",
    "df_meta.set_index('Cell', inplace=True)\n",
    "df_meta.to_csv('cellphoneDB/input/meta.tsv', sep = '\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Run cellphoneDB\n",
    "\n",
    "```\n",
    "source ~/cpdb-venv/bin/activate\n",
    "```\n",
    "\n",
    "## Create database\n",
    "```\n",
    "cellphonedb database generate \\\n",
    "    --result-path database \\\n",
    "    --user-gene ~/gonads/scRNAseq_analysis_FCA_2sex/cellphoneDB/database/gene_input_all.csv \\\n",
    "    --user-complex ~/gonads/scRNAseq_analysis_FCA_2sex/cellphoneDB/database/complex_curated_032021.tsv \\\n",
    "    --user-interactions ~/gonads/scRNAseq_analysis_FCA_2sex/cellphoneDB/database/interaction_curated_032021.tsv\n",
    "```\n",
    "\n",
    "## Run cellphone without statistical analysis - just get expressed interactions\n",
    "```\n",
    "cellphonedb method analysis \\\n",
    "    ~/gonads/main/cellphoneDB/input/meta.tsv \\\n",
    "    ~/gonads/main/cellphoneDB/input/counts.h5ad \\\n",
    "    --database ~/gonads/main/cellphoneDB/database/cellphonedb_user_2021-03-11-15_06.db \\\n",
    "    --counts-data hgnc_symbol \\\n",
    "    --output-path ~/gonads/main/cellphoneDB/out/ \\\n",
    "    --threshold 0\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Merge DEGs files\n",
    "\n",
    "```\n",
    "cp germcells_DEGs.csv merged_DEGs.csv \\\n",
    "grep gene -v supporting_DEGs.csv >> merged_DEGs.csv \\\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
