{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import anndata as an\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "# import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import os\n",
    "\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "plt.rcParams[\"font.family\"] = \"Arial\"\n",
    "\n",
    "sc.set_figure_params(figsize=(4, 4))\n",
    "\n",
    "palette = ['#199de5', '#fbbc04']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "sc.set_figure_params(facecolor=\"white\", figsize=(8, 8), dpi_save=300)\n",
    "sc.settings.verbosity = 1\n",
    "sc.settings.figdir = './figures-visium/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "path = '/nfs/team292/lg18/gonads/data/visium/'\n",
    "\n",
    "\n",
    "# sample IDs\n",
    "donors_match = {\n",
    "           'Hrv15':[\"F94\", \"Hrv15\", \"Hrv18\", 'Hrv32', 'Hrv10', 'F88', 'Hrv27', 'Hrv25'],\n",
    "           'Hrv69':['Hr69', 'Hrv58', 'Hrv59', 'Hrv65'],\n",
    "           'F83':[\"F83\", \"Hrv34\",'Hrv86', 'F133','Hrv59'], \n",
    "           'F91':[\"F91\", \"F87\", 'Hrv37', 'Hrv30','Hrv26'],\n",
    "           'F94':[\"F94\", \"Hrv15\", \"Hrv18\", 'Hrv32', 'Hrv10', 'F88'],\n",
    "           'Hrv11':[\"Hrv11\", \"Hrv85\", \"Hrv91\", \"F96\", \"Hrv39\",\"Hrv57\"],\n",
    "           'F81':['F81', 'Hrv12', 'Hrv47', 'F99', 'F126','F84','Hrv13'],\n",
    "           'Hrv41':['Hrv41', 'Hrv31'],\n",
    "           'Hrv58':['Hrv58','Hr69', 'Hrv59', 'Hrv65'],\n",
    "           'Hrv3':['Hrv3', 'Hrv45', 'Hrv51', 'Hrv19','Hrv71', 'Hrv54'],\n",
    "           'Hrv27':['Hrv27', 'Hrv25', 'Hrv35', 'F122', 'Hrv40', 'Hrv33'],\n",
    "           'Hrv13':['Hrv13', 'F99', 'F126','F84']}\n",
    "\n",
    "Vsamples = {'F91':[\"FCA_GNDST8796127\", \"FCA_GNDST8796128\"],\n",
    "           'Hrv11':[\"FCA_GNDST8796131\", \"FCA_GNDST8796132\"],\n",
    "           'F83':[\"FCA_GNDST8796125\", \"FCA_GNDST8796126\"], \n",
    "           'F94':[\"FCA_GNDST8796129\", \"FCA_GNDST8796130\"],\n",
    "           'F81':['FCA_GNDST9421027'],\n",
    "           'Hrv41':['FCA_GNDST9421026'],\n",
    "           'Hrv69':['FCA_GNDST10484130', 'FCA_GNDST10484131'],\n",
    "           'Hrv58':['FCA_GNDST10484129','FCA_GNDST10484132'],\n",
    "           'Hrv3':['FCA_GNDST10485488'],\n",
    "           'Hrv27':['FCA_GNDST10485487'],\n",
    "           'Hrv13':['FCA_GNDST10485485'],\n",
    "           'Hrv15':['FCA_GNDST10485486'],\n",
    "           'females':[\"FCA_GNDST8796125\", \"FCA_GNDST8796126\", \n",
    "                       \"FCA_GNDST8796127\", \"FCA_GNDST8796128\", \n",
    "                       \"FCA_GNDST8796131\", \"FCA_GNDST8796132\",\n",
    "                      'FCA_GNDST9421027',\n",
    "                      'FCA_GNDST10484130', 'FCA_GNDST10484131',\n",
    "                       'FCA_GNDST10484129','FCA_GNDST10484132'],\n",
    "            'males':[\"FCA_GNDST8796129\", \"FCA_GNDST8796130\",\n",
    "                     'FCA_GNDST9421026', \n",
    "                     'FCA_GNDST10485488',\n",
    "                     'FCA_GNDST10485487',\n",
    "                     'FCA_GNDST10485485',\n",
    "                     'FCA_GNDST10485486']\n",
    "           }\n",
    "            "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reading Visium data\n",
    "\n",
    "\n",
    "The function read_visium returns an AnnData object that contains counts, images and spatial coordinates. We will calculate standards QC metrics with pp.calculate_qc_metrics and visualize them.\n",
    "\n",
    "An anndata will be saved for cell2location.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_and_qc(sample_name, path):\n",
    "\n",
    "    adata = sc.read_visium(path + str(sample_name),\n",
    "                           count_file='filtered_feature_bc_matrix.h5', load_images=True)\n",
    "    adata.obs['sample'] = sample_name\n",
    "    adata.var['SYMBOL'] = adata.var_names\n",
    "    adata.var.rename(columns={'gene_ids': 'ENSEMBL'}, inplace=True)\n",
    "    adata.var_names = adata.var['ENSEMBL']\n",
    "    adata.var.drop(columns='ENSEMBL', inplace=True)\n",
    "\n",
    "    # Calculate QC metrics\n",
    "    sc.pp.calculate_qc_metrics(adata, inplace=True)\n",
    "    adata.var['mt'] = [gene.startswith('mt-') for gene in adata.var['SYMBOL']]\n",
    "    adata.obs['mt_frac'] = adata[:, adata.var['mt'].tolist()].X.sum(1).A.squeeze()/adata.obs['total_counts']\n",
    "\n",
    "    # mitochondria-encoded (MT) genes should be removed for spatial mapping\n",
    "    adata.obsm['mt'] = adata[:, adata.var['mt'].values].X.toarray()\n",
    "    adata = adata[:, ~adata.var['mt'].values]\n",
    "\n",
    "    # add sample name to obs names\n",
    "    adata.obs[\"sample\"] = [str(i) for i in adata.obs['sample']]\n",
    "    adata.obs_names = adata.obs[\"sample\"] \\\n",
    "                          + '_' + adata.obs_names\n",
    "    adata.obs.index.name = 'spot_id'\n",
    "\n",
    "    return adata\n",
    "\n",
    "\n",
    "def select_slide(adata, s, s_col='sample'):\n",
    "    r\"\"\" Select data for one slide from the spatial anndata object.\n",
    "\n",
    "    :param adata: Anndata object with multiple spatial samples\n",
    "    :param s: name of selected sample\n",
    "    :param s_col: column in adata.obs listing sample name for each location\n",
    "    \"\"\"\n",
    "\n",
    "    slide = adata[adata.obs[s_col].isin([s]), :]\n",
    "    s_keys = list(slide.uns['spatial'].keys())\n",
    "    s_spatial = np.array(s_keys)[[s in k for k in s_keys]][0]\n",
    "\n",
    "    slide.uns['spatial'] = {s_spatial: slide.uns['spatial'][s_spatial]}\n",
    "\n",
    "    return slide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "F91\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "F83\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "F94\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "F81\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv41\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv69\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv58\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv27\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "females\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10484129' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10484130' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10484131' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10484132' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796125' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796126' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796127' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796128' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796131' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796132' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST9421027' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "males\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "... storing 'sample_name' as categorical\n",
      "... storing 'feature_types' as categorical\n",
      "... storing 'genome' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10485485' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10485486' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10485487' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST10485488' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796129' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST8796130' as categorical\n",
      "... storing 'SYMBOL-FCA_GNDST9421026' as categorical\n"
     ]
    }
   ],
   "source": [
    "for sam in Vsamples.keys():\n",
    "    print(sam)\n",
    "    # read first sample\n",
    "    adata = read_and_qc(Vsamples[sam][0], path=path)\n",
    "\n",
    "    # read the remaining samples\n",
    "    slides = {}\n",
    "    for i in Vsamples[sam][1:]:\n",
    "        adata_1 = read_and_qc(i, path=path)\n",
    "        slides[str(i)] = adata_1\n",
    "\n",
    "    adata_0 = adata.copy()\n",
    "\n",
    "    # combine individual samples\n",
    "    adata = adata.concatenate(\n",
    "        list(slides.values()),\n",
    "        batch_key=\"sample\",\n",
    "        uns_merge=\"unique\",\n",
    "        batch_categories=Vsamples[sam],\n",
    "        index_unique=None\n",
    "    )\n",
    "    # merging metadata\n",
    "    adata.obs['sample_name'] = sam\n",
    "\n",
    "    \n",
    "    os.system('mkdir -p '+ path+'/cell2location/'+ sam +'/')  \n",
    "    adata.write(path + '/cell2location/' + sam + '/'+ sam + '_visium.h5ad')\n",
    "      "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# VsamplesPaper  = {'F91':[\"FCA_GNDST8796127\", \"FCA_GNDST8796128\"],\n",
    "#            'Hrv11':[\"FCA_GNDST8796131\", \"FCA_GNDST8796132\"],\n",
    "#            'F83':[\"FCA_GNDST8796125\", \"FCA_GNDST8796126\"], \n",
    "#            'F94':[\"FCA_GNDST8796129\", \"FCA_GNDST8796130\"],\n",
    "#                    'Hr69':['FCA_GNDST10484130', 'FCA_GNDST10484131'],\n",
    "#             'Hrv58':['FCA_GNDST10484129','FCA_GNDST10484132'],\n",
    "#                   'Hrv27':['FCA_GNDST10485487'],\n",
    "#              'Hrv3':['FCA_GNDST10485488']\n",
    "#                  }\n",
    "\n",
    "\n",
    "# fig, axs = plt.subplots(4, 2, figsize=(8, 15))\n",
    "\n",
    "# j = 0\n",
    "# for sam in VsamplesPaper.keys():\n",
    "#     i = 0\n",
    "#     for sli in Vsamples[sam]:\n",
    "#         spt_adata =  read_and_qc(Vsamples[sam][0], path=path)\n",
    "#         sns.distplot(spt_adata.obs[\"n_genes_by_counts\"], kde=False, bins=100, hist_kws={\"alpha\": 0.2}, ax=axs[j][i])\n",
    "#         i = i + 1\n",
    "#     j = j + 1\n",
    "      "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Reading the merged scRNA dataset\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(396565, 28820)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "path2sc = '/nfs/team292/lg18/with_valentina/gonadsV2_revision/'\n",
    "adata = sc.read(path2sc+'FCA-gonads_rawcounts.h5ad')\n",
    "adata.X.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (29,30,32) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (29,30) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    }
   ],
   "source": [
    "# Add old cluster subname\n",
    "metaF = pd.read_csv(path2sc+'/FCA-gonads_female_annotated.csv')\n",
    "metaF = metaF.set_index('Unnamed: 0')\n",
    "metaF.head()\n",
    "\n",
    "metaM = pd.read_csv(path2sc+'/FCA-gonads_male_annotated.csv')\n",
    "metaM = metaM.set_index('Unnamed: 0')\n",
    "metaM.head()\n",
    "\n",
    "\n",
    "cell_annotM = metaM['lineage'].to_dict()\n",
    "cell_annotF = metaF['lineage'].to_dict()\n",
    "cell_annot = {**cell_annotM , **cell_annotF}\n",
    "adata.obs['lineage'] = adata.obs_names.map(cell_annot).astype('category')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Add cell type annotations from subanalysis as LABELS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (29,30,37) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/IPython/core/interactiveshell.py:3166: DtypeWarning: Columns (29,30,31,32,37) have mixed types.Specify dtype option on import or set low_memory=False.\n",
      "  interactivity=interactivity, compiler=compiler, result=result)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Mesenchymal_GATA2       84229\n",
       "Sertoli                 36917\n",
       "lowQC                   35077\n",
       "Ti                      34808\n",
       "Oi                      25730\n",
       "preGC_IIb               23825\n",
       "Epithelial              20222\n",
       "preGC_I                 15599\n",
       "Endothelial             14484\n",
       "preGC_IIa               13772\n",
       "PV                      12679\n",
       "OSE                      8222\n",
       "Doublet                  6091\n",
       "early_supporting         5894\n",
       "CoelEpi_LHX9             5547\n",
       "Immune                   5533\n",
       "Gi                       4749\n",
       "cont                     4145\n",
       "CoelEpi_GATA2            4050\n",
       "Neural                   3783\n",
       "PGC                      3764\n",
       "early_sPAX8              3648\n",
       "GermCells                3523\n",
       "SMCs                     3044\n",
       "FetalLeydig              2874\n",
       "oogonia_STRA8            2203\n",
       "late_sPAX8               2118\n",
       "oogonia_meiotic          1875\n",
       "CoelEpi_LHX9_cycling     1858\n",
       "early_somatic            1414\n",
       "Erythroid                1371\n",
       "pre_oocyte                812\n",
       "oocyte                    723\n",
       "pre_spermatogonia         621\n",
       "GC_mitotic                510\n",
       "GC                        485\n",
       "granulosa                 366\n",
       "Name: celltype, dtype: int64"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add  cluster subname\n",
    "supF = pd.read_csv(path2sc+'/FCA-gonads_somatic_annotated.csv')\n",
    "supF = supF.set_index('Unnamed: 0')\n",
    "supF.head()\n",
    "\n",
    "supM = pd.read_csv(path2sc+'/FCA-gonads_somaticEarly_annotated.csv')\n",
    "supM = supM.set_index('Unnamed: 0')\n",
    "supM.head()\n",
    "\n",
    "germs = pd.read_csv(path2sc+'/FCA-gonads_germcellsClean_annotated.csv')\n",
    "germs = germs.set_index('Unnamed: 0')\n",
    "germs.head()\n",
    "\n",
    "cell_annotSM = supM['celltype'].to_dict()\n",
    "cell_annotSF = supF['celltype'].to_dict()\n",
    "cell_annotG = germs['celltype'].to_dict()\n",
    "cell_annot = {**cell_annotSM , **cell_annotSF, **cell_annotG}\n",
    "adata.obs['celltype'] = adata.obs_names.map(cell_annot)#.astype('category')\n",
    "\n",
    "\n",
    "# Replace NaN with big clusters \n",
    "adata.obs['celltype'] = adata.obs['celltype'].fillna(adata.obs['lineage'])\n",
    "adata.obs['labels'] = adata.obs['celltype']\n",
    "\n",
    "adata.obs['celltype'].value_counts(dropna = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Mesenchymal_GATA2    84229\n",
       "Sertoli              36917\n",
       "Ti                   34808\n",
       "Oi                   25730\n",
       "preGC_IIb            23825\n",
       "Epithelial           20222\n",
       "preGC_I              15599\n",
       "Endothelial          14484\n",
       "preGC_IIa            13772\n",
       "PV                   12679\n",
       "OSE                   8222\n",
       "early_supporting      5894\n",
       "CoelEpi_LHX9          5547\n",
       "Immune                5533\n",
       "Gi                    4749\n",
       "CoelEpi_GATA2         4050\n",
       "Neural                3783\n",
       "PGC                   3764\n",
       "early_sPAX8           3648\n",
       "SMCs                  3044\n",
       "FetalLeydig           2874\n",
       "oogonia_STRA8         2203\n",
       "late_sPAX8            2118\n",
       "oogonia_meiotic       1875\n",
       "early_somatic         1414\n",
       "pre_oocyte             812\n",
       "oocyte                 723\n",
       "pre_spermatogonia      621\n",
       "GC_mitotic             510\n",
       "GC                     485\n",
       "granulosa              366\n",
       "Name: labels, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## Remove unreliable populations\n",
    "# adata = adata[[ 'G1' in i for i in adata.obs.phase]]\n",
    "adata = adata[[ 'cycling' not in i for i in adata.obs.labels]]\n",
    "adata = adata[[ i not in ['GermCells', 'Erythroid', 'cont', 'lowQC', 'Doublet'] for i in adata.obs.labels]]\n",
    "set(adata.obs.labels)\n",
    "adata.obs.labels.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Use Ensembl id as GENE "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.var` of view, copying.\n"
     ]
    }
   ],
   "source": [
    "adata.var['SYMBOL'] = adata.var_names\n",
    "adata.var.rename(columns={'gene_ids-0': 'ENSEMBL'}, inplace=True)\n",
    "adata.var_names = adata.var['ENSEMBL']\n",
    "adata.var.drop(columns='ENSEMBL', inplace=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Slice dataset per individual"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Hrv15\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9           534\n",
      "Endothelial           3044\n",
      "Epithelial             294\n",
      "FetalLeydig            753\n",
      "Immune                 868\n",
      "Mesenchymal_GATA2    15147\n",
      "Neural                  56\n",
      "PGC                    114\n",
      "PV                    1137\n",
      "Sertoli              12982\n",
      "Ti                   13314\n",
      "late_sPAX8            1644\n",
      "pre_spermatogonia        4\n",
      "dtype: int64\n",
      "Hrv69\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9           21\n",
      "Endothelial          1692\n",
      "Epithelial           4096\n",
      "GC_mitotic             32\n",
      "Gi                     26\n",
      "Immune                570\n",
      "Mesenchymal_GATA2    8791\n",
      "Neural                415\n",
      "OSE                   104\n",
      "Oi                   2595\n",
      "PGC                    49\n",
      "PV                   4479\n",
      "SMCs                 3031\n",
      "granulosa             339\n",
      "oocyte                225\n",
      "oogonia_STRA8         145\n",
      "oogonia_meiotic       235\n",
      "preGC_I               166\n",
      "preGC_IIa             116\n",
      "preGC_IIb             822\n",
      "pre_oocyte            170\n",
      "dtype: int64\n",
      "F83\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9           202\n",
      "Endothelial           1932\n",
      "GC                     217\n",
      "GC_mitotic             304\n",
      "Gi                     280\n",
      "Immune                 401\n",
      "Mesenchymal_GATA2     6301\n",
      "Neural                  82\n",
      "OSE                   3820\n",
      "Oi                    9809\n",
      "PGC                   1311\n",
      "PV                    1322\n",
      "early_sPAX8             35\n",
      "granulosa               26\n",
      "oocyte                 525\n",
      "oogonia_STRA8         1489\n",
      "oogonia_meiotic       1506\n",
      "preGC_I               2286\n",
      "preGC_IIa             2639\n",
      "preGC_IIb            12416\n",
      "pre_oocyte             646\n",
      "dtype: int64\n",
      "F91\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_GATA2          96\n",
      "CoelEpi_LHX9          409\n",
      "Endothelial           654\n",
      "Epithelial            711\n",
      "GC                     20\n",
      "GC_mitotic             55\n",
      "Gi                    333\n",
      "Immune                299\n",
      "Mesenchymal_GATA2    6470\n",
      "Neural                 45\n",
      "OSE                  1826\n",
      "Oi                   4417\n",
      "PGC                   250\n",
      "PV                    386\n",
      "early_sPAX8           305\n",
      "early_supporting       29\n",
      "oogonia_STRA8         257\n",
      "oogonia_meiotic        49\n",
      "preGC_I              5558\n",
      "preGC_IIa            4347\n",
      "preGC_IIb            5245\n",
      "pre_oocyte             10\n",
      "dtype: int64\n",
      "F94\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9           474\n",
      "Endothelial           1371\n",
      "Epithelial             292\n",
      "FetalLeydig            394\n",
      "Immune                 602\n",
      "Mesenchymal_GATA2    12443\n",
      "Neural                  51\n",
      "PGC                    104\n",
      "PV                     654\n",
      "Sertoli              10744\n",
      "Ti                    5450\n",
      "late_sPAX8             826\n",
      "dtype: int64\n",
      "Hrv11\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9          237\n",
      "Endothelial          1239\n",
      "Epithelial           2404\n",
      "GC                     29\n",
      "GC_mitotic             73\n",
      "Gi                    110\n",
      "Immune                217\n",
      "Mesenchymal_GATA2    4075\n",
      "Neural                623\n",
      "OSE                   910\n",
      "Oi                   2591\n",
      "PGC                   277\n",
      "PV                   3070\n",
      "early_sPAX8            32\n",
      "oocyte                 13\n",
      "oogonia_STRA8         222\n",
      "oogonia_meiotic       163\n",
      "preGC_I               823\n",
      "preGC_IIa            1192\n",
      "preGC_IIb            4238\n",
      "pre_oocyte             63\n",
      "dtype: int64\n",
      "F81\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_GATA2         171\n",
      "CoelEpi_LHX9          628\n",
      "Endothelial           417\n",
      "Epithelial            632\n",
      "Gi                    342\n",
      "Immune                125\n",
      "Mesenchymal_GATA2    7968\n",
      "Neural                257\n",
      "OSE                   943\n",
      "Oi                   3967\n",
      "PGC                   187\n",
      "PV                    239\n",
      "early_sPAX8           957\n",
      "early_somatic          38\n",
      "early_supporting      143\n",
      "preGC_I              5540\n",
      "preGC_IIa            4674\n",
      "dtype: int64\n",
      "Hrv41\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_GATA2         510\n",
      "CoelEpi_LHX9          411\n",
      "Endothelial           518\n",
      "Epithelial            860\n",
      "FetalLeydig           200\n",
      "Immune                384\n",
      "Mesenchymal_GATA2    3724\n",
      "PGC                    12\n",
      "PV                    118\n",
      "Sertoli               883\n",
      "Ti                    605\n",
      "early_sPAX8           348\n",
      "late_sPAX8            117\n",
      "dtype: int64\n",
      "Hrv58\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9           21\n",
      "Endothelial          1692\n",
      "Epithelial           4096\n",
      "GC_mitotic             32\n",
      "Gi                     26\n",
      "Immune                570\n",
      "Mesenchymal_GATA2    8791\n",
      "Neural                415\n",
      "OSE                   104\n",
      "Oi                   2595\n",
      "PGC                    49\n",
      "PV                   4479\n",
      "SMCs                 3031\n",
      "granulosa             339\n",
      "oocyte                225\n",
      "oogonia_STRA8         145\n",
      "oogonia_meiotic       235\n",
      "preGC_I               166\n",
      "preGC_IIa             116\n",
      "preGC_IIb             822\n",
      "pre_oocyte            170\n",
      "dtype: int64\n",
      "Hrv3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9           93\n",
      "Endothelial           339\n",
      "Epithelial            188\n",
      "FetalLeydig           348\n",
      "GC                    101\n",
      "GC_mitotic             10\n",
      "Immune                358\n",
      "Mesenchymal_GATA2    1241\n",
      "PGC                   340\n",
      "PV                    226\n",
      "Sertoli              5515\n",
      "Ti                   3242\n",
      "late_sPAX8             70\n",
      "pre_spermatogonia     390\n",
      "dtype: int64\n",
      "Hrv27\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_LHX9           674\n",
      "Endothelial           2236\n",
      "Epithelial             308\n",
      "FetalLeydig            799\n",
      "GC                      91\n",
      "GC_mitotic              31\n",
      "Immune                 850\n",
      "Mesenchymal_GATA2     9494\n",
      "Neural                  22\n",
      "PGC                   1003\n",
      "PV                     737\n",
      "Sertoli              13313\n",
      "Ti                   18125\n",
      "late_sPAX8             939\n",
      "pre_spermatogonia      115\n",
      "dtype: int64\n",
      "Hrv13\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:119: ImplicitModificationWarning: Transforming to str index.\n",
      "  warnings.warn(\"Transforming to str index.\", ImplicitModificationWarning)\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/pandas/core/arrays/categorical.py:2487: FutureWarning: The `inplace` parameter in pandas.Categorical.remove_unused_categories is deprecated and will be removed in a future version.\n",
      "  res = method(*args, **kwargs)\n",
      "/home/jovyan/my-conda-envs/cellpymc/lib/python3.7/site-packages/anndata/_core/anndata.py:1208: ImplicitModificationWarning: Initializing view as actual.\n",
      "  \"Initializing view as actual.\", ImplicitModificationWarning\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'celltype' as categorical\n",
      "Trying to set attribute `.obs` of view, copying.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "Variable names are not unique. To make them unique, call `.var_names_make_unique`.\n",
      "... storing 'labels' as categorical\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "labels\n",
      "CoelEpi_GATA2         118\n",
      "CoelEpi_LHX9          504\n",
      "Endothelial           178\n",
      "Epithelial            484\n",
      "Gi                    136\n",
      "Immune                 68\n",
      "Mesenchymal_GATA2    2723\n",
      "Neural                 17\n",
      "OSE                   598\n",
      "Oi                   1789\n",
      "PGC                   173\n",
      "PV                    120\n",
      "early_sPAX8           663\n",
      "early_somatic          36\n",
      "early_supporting      109\n",
      "preGC_I              2759\n",
      "preGC_IIa            1797\n",
      "preGC_IIb              36\n",
      "dtype: int64\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAiMAAAINCAYAAADsjH/3AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAxOAAAMTgF/d4wjAAAkyklEQVR4nO3db4xV9Z348c9MYRhxVYYIiIKhO4DAVErgB9gnRm0mVoS6pDHqCml3jSSrUTegwT+BuXY3K0J3VmJ5UFO323UNWnBRAVM2VKpulWpttlOY0a4oWlJw/TOL2gozHc7vQcvU2YHhzjD3fJk7r1dCzD33nnO/fDp63z33zrkVWZZlAQCQSGXqBQAAg5sYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQ1JPUCTmTYsGExatSokhy7vb09hg4dWpJj05VZ58es82HO+THr/JRy1u+9914cPnz4mPed8jEyatSo2LdvX0mO3dTUFNOnTy/JsenKrPNj1vkw5/yYdX5KOetx48Yd9z5v0wAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkNST1AlKavHBhRHV1aQ6+e3dpjgsAZcaZkVLYsyeiri71KgBgQBAjAEBSg/ptmqpf/zqisgQ9dvjwH86OAAAnNKhjpG38+KguxWdGmpsj2tr6/7gAUIYGdYz8atOmmD59ev8fuBRnWwCgTHnVBACSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACCpomJk48aNsXDhwjj//PNj+PDhMW3atFi9enW0tbX1uF+hUIiKiopuf26//fZ+WTwAMPANKeZB3/rWt2LChAmxevXqOOecc+LFF1+MhoaG+OUvfxmPPPJIj/tWVVXFc88912Xbueee2/cV96PJCxdGVFf3/4GzLKKiov+PCwBlqKgY2bx5c4waNarz9iWXXBJHjhyJFStWxJo1a+Kcc8457r4VFRVx0UUXnfxKB5osS70CABgQioqRz4bIUbNmzYqIiP379/cYI6eyX23aFNOnT+//AzsrAgBFKypGjuWFF16IqqqqqK2t7fFxbW1tMWbMmPjggw9iwoQJsWTJkrj99tujsjL9Z2dL9jYNAFC0PsVIS0tLrF27NpYsWRJnnnnmcR83ceLEuP/++2PGjBnR0dERTz75ZNx5553x9ttvx7p16/q8aACgfFRkWe8+3PDBBx/El770paiuro6XXnopTj/99F494bJly+KBBx6Id955J84777xu9zc2NkZjY2Pn7YMHD8ZPfvKTXj1Hsdrb22Po0KH9ftwLv/jFiIj45S9+0e/HHqhKNWu6M+t8mHN+zDo/pZz1vHnzYt++fce8r1cx8sknn8SXv/zlePfdd+PFF1/s02/FvPzyyzF37tx4+umnY8GCBSd8/Lhx4467+JPV1NRU2s+M+BBrp5LNmm7MOh/mnB+zzk8pZ93T63nRb9McPnw4Fi5cGG+++Wb853/+50n/em6FD3kCAFFkjHR0dMR1110XO3fujGeffTYuuOCCPj/h+vXro7KyMmbOnNnnYwAA5aOoGLn55ptj06ZN8Xd/93fR0dERO3fu7LyvtrY2Ro0aFYVCIe6999546623YsKECRERMXPmzPj6178eU6ZMifb29njqqafi4YcfjltvvfWUufAZAJBWUTHywx/+MCIiVqxYEStWrOhy3/e+9734xje+ccz9Jk2aFGvXro39+/dHRMSUKVPiwQcfjJtuuukklgwAlJOiYmTv3r0nfEyhUIhCodBl2+OPP96XNQEAg0ifL3pWDr7w//5fxClw8TUAGMy8EgMASQ3qMyNt48dHdSkuB9/c3P/HBIAy5cwIAJDUoD4z4lt7ASA9Z0YAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkNSgvhz85IULI0rxRXkAQNGcGQEAkhrUZ0Z8UR4ApOfMCACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgqSGpF5DS5IULI6qrUy8DAAa1QR0jVb/+dUSlk0MAkNKgjpG28eOjuhRnRpqb+/+YAFCmBnWM/GrTppg+fXr/H7iiov+PCQBlynsUAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQVFExsnHjxli4cGGcf/75MXz48Jg2bVqsXr062traTrjv9u3bY/bs2VFdXR3jx4+PhoaG6OjoOOmFAwDloagY+da3vhXDhg2L1atXxzPPPBOLFi2KhoaGuOGGG3rc75VXXol58+bF5MmT45lnnonly5fHmjVrYvny5f2yeABg4CvqCqybN2+OUaNGdd6+5JJL4siRI7FixYpYs2ZNnHPOOcfcr1AoRF1dXTzyyCNRWVkZl112WXz66adxzz33xB133BFjxozpn78FADBgFXVm5LMhctSsWbMiImL//v3H3KetrS22b98e1157bVR+5svoFi1aFO3t7bFt27a+rBcAKDN9/gDrCy+8EFVVVVFbW3vM+/fs2RNtbW1RV1fXZfvYsWOjpqYmWlpa+vrUAEAZ6VOMtLS0xNq1a2PJkiVx5plnHvMxra2tERExYsSIbveNHDkyPvzww748NQBQZnr9rb0ffPBBXHXVVVFbWxurVq067uOyLIuIiIpjfIPt0fuOpbGxMRobGztvHzx4MJqamnq7zKK0t7eX5NgX/vGfvyzRugeiUs2a7sw6H+acH7POT6pZ9ypGPvnkk5g3b160tbXFj3/84zj99NOP+9iRI0dGxJ/OkHxWa2tr5/3/19KlS2Pp0qWdt8eNGxfTp0/vzTKL1tTUVLJjR0RJjz3QlHrW/IlZ58Oc82PW+Uk166Lfpjl8+HAsXLgw3nzzzdi2bVuce+65PT6+trY2qqqqorm5ucv2AwcORGtra0ydOrVvKwYAykpRMdLR0RHXXXdd7Ny5M5555pm44IILTrhPVVVV1NfXx2OPPdblbZlHH300hg4dGpdffnnfVw0AlI2iYuTmm2+OTZs2xfLly6OjoyN27tzZ+ee9996LiD9cU6SioiL27t3bud/KlStj165dsXjx4tixY0esW7cuVq5cGbfeeqtrjAAAEVFkjPzwhz+MiIgVK1bEl770pS5/tm7detz95syZE1u3bo3XXnstrrjiirjvvvti2bJlPX7wFQAYXIr6AOtnz3YcT6FQiEKh0G17fX191NfX93ZdAMAg4Vt7AYCkxAgAkFSvL3pWTiYvXBhRXZ16GQAwqDkzAgAkNajPjPxq06bSXGnuGJfABwCOzZkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSRcfIvn374pZbbom5c+dGdXV1VFRUFLVfoVCIioqKbn9uv/32Pi8aACgfQ4p94BtvvBEbNmyI2bNnx9y5c+P5558v+kmqqqriueee67Lt3HPPLX6VAEDZKjpGLr744jhw4EBERKxatapXMVJRUREXXXRR71cHAJS9ot+mqaz08RIAoP/lUhhtbW0xZsyYGDJkSEycODFWr14dR44cyeOpAYBTXNFv0/TVxIkT4/77748ZM2ZER0dHPPnkk3HnnXfG22+/HevWrSv10wMAp7iKLMuy3u60atWquOuuu6IPu0ZExLJly+KBBx6Id955J84777wu9zU2NkZjY2Pn7YMHD8ZPfvKTPj3PibS3t8fQoUP7/bgXfvGLERHxy1/8ot+PPVCVatZ0Z9b5MOf8mHV+SjnrefPmxb59+455X5IYefnll2Pu3Lnx9NNPx4IFC3p87Lhx4467+JPV1NQU06dP7/8DH/215z7OpxyVbNZ0Y9b5MOf8mHV+Sjnrnl7Pk34qtdhrlQAA5StJjKxfvz4qKytj5syZKZ4eADiF9OoDrBs3boyIiF27dnW5PW3atJg2bVoUCoW4995746233ooJEyZERMTMmTPj61//ekyZMiXa29vjqaeeiocffjhuvfVWFz4DAHoXI1dfffUxbzc0NEShUDjmPpMmTYq1a9fG/v37IyJiypQp8eCDD8ZNN93Uh+UCAOWmVzFyog+sFgqFblHy+OOP93pRAMDg4bKqAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEkVHSP79u2LW265JebOnRvV1dVRUVFR9JNs3749Zs+eHdXV1TF+/PhoaGiIjo6OPi0YACgvRcfIG2+8ERs2bIjRo0fH3Llzi36CV155JebNmxeTJ0+OZ555JpYvXx5r1qyJ5cuX92nBAEB5GVLsAy+++OI4cOBARESsWrUqnn/++aL2KxQKUVdXF4888khUVlbGZZddFp9++mncc889cccdd8SYMWP6tnIAoCwUfWaksrL3Hy9pa2uL7du3x7XXXttl/0WLFkV7e3ts27at18cEAMpLST/AumfPnmhra4u6urou28eOHRs1NTXR0tJSyqcHAAaAksZIa2trRESMGDGi230jR46MDz/8sJRPDwAMAEV/ZqQvsiyLiDjmb94cve//amxsjMbGxs7bBw8ejKamppKsr729vSTHvvCP//xlidY9EJVq1nRn1vkw5/yYdX5SzbqkMTJy5MiI+NMZks9qbW3tvP+zli5dGkuXLu28PW7cuJg+fXpJ1tfU1FSyY0dESY890JR61vyJWefDnPNj1vlJNeuSvk1TW1sbVVVV0dzc3GX7gQMHorW1NaZOnVrKpwcABoCSxkhVVVXU19fHY4891uVtmUcffTSGDh0al19+eSmfHgAYAHoVIxs3boyNGzfGrl27utw+euajUChERUVF7N27t3OflStXxq5du2Lx4sWxY8eOWLduXaxcuTJuvfVW1xgBAHr3mZGrr776mLcbGhqiUCgcc585c+bE1q1b46677oorrrgizj777Fi2bFmsXLmybysGAMpKr2LkeL8Bc1ShUDhmlNTX10d9fX2vFgYADA6+tRcASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJDUi+grNXVpV7BKWPyoUMR1dWplzEomHU+zDk/Zp2j9euTPK0zIwBAUmIEAEhKjAAASYkRACApMQIAJOW3aUpp9+7UKzhl/KqpKaZPn556GYOCWefDnPNj1jlqakrytM6MAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApIqOkTfffDMWLFgQZ5xxRtTU1MTixYvj/fffP+F+hUIhKioquv25/fbbT2rhAEB5GFLMgz766KO49NJLY9SoUbF+/fr49NNP484774wrr7wyXnrppais7Llpqqqq4rnnnuuy7dxzz+37qgeKurrUKzhlTD50KKK6OvUyBgWzzoc558esc7R+fZKnLSpGvvOd78S7774bO3fujLFjx0ZExIQJE2LOnDnx1FNPxcKFC3vcv6KiIi666KKTXy0AUHaKipGtW7fGpZde2hkiERGzZ8+OyZMnx+bNm08YI4PW7t2pV3DK+FVTU0yfPj31MgYFs86HOefHrHPU1JTkaYv6zEhLS0vUHeMth7q6umhpaTnh/m1tbTFmzJgYMmRITJw4MVavXh1Hjhzp/WoBgLJT1JmR1tbWGDFiRLftI0eOjN0n+H//EydOjPvvvz9mzJgRHR0d8eSTT8add94Zb7/9dqxbt65Pix4wfGakk/d882PW+TDn/Jh1jk7lz4xE/OFzH/9XlmUn3G/RokVdbn/lK1+J008/PR544IG4++6747zzzutyf2NjYzQ2NnbePnjwYDSV6LRRe3t7SY594R//efjQoX4/9kCVRcQh88iFWefDnPNj1vkp1eviiRQVIzU1NdHa2tpte2tra4wcObLXT3rNNddEY2Nj/PznP+8WI0uXLo2lS5d23h43blzJ3itsKvH7kNV79pTs2ANNqWfNn5h1Psw5P2adn6GJZl3UZ0amTp0azc3N3bY3NzfH1KlT+/zkxzrbAgAMLkXFyPz582PHjh1x4MCBzm2vvvpqvP7667FgwYJeP+n69eujsrIyZs6c2et9AYDyUlSMLFmyJEaPHh1f/epXY+vWrfHEE0/ENddcE3PmzImrrrqq83FHr7a6d+/ezm0zZ86MtWvXxrZt22LLli1x4403xtq1a+OWW24ZHBc+AwB6VNRnRs4888x49tln47bbbotrrrkmhgwZEvPnz49/+qd/OuHVVydNmhRr166N/fv3R0TElClT4sEHH4ybbrrp5FcPAAx4Rf82TW1tbWzZsqXHxxQKhSgUCl22Pf74431aGAAwOBQdI/SB34vv9IUjRyJOcBaN/mHW+TDn/Jh1jl5+OcnT+l8XAEjKmZFScpGeTrtcJyA3Zp0Pc86PWeco0XfTiJFScjn4Ti7nnB+zzoc558esc5TocvDepgEAknJmpJRO8CWCg4mvAM+PWefDnPNj1jlK9DaNMyMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASMrl4EvJF+V18kVX+THrfJhzfsw6R74or4xUVKReAQAMGM6MlEpFhS/K+wxfdJUfs86HOefHrHPki/IAgMFIjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJiREAICkxAgAkJUYAgKTECACQlBgBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJDUkNQLKFtZFlFXl3oVp4zJhw5FVFenXsagYNb5MOf8mHWO1q9P8rRipBSqqiIOH47Ysyf1Sk4ZVUeORFQ6EZcHs86HOefHrMufGCmF2to/hEhtbeqVnDLaDh2Kav/PJhdmnQ9zzo9Zlz8xUiq1tRG7d6dexSnjV01NMX369NTLGBTMOh/mnB+zzlFTU5Kndd4LAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSKjpG3nzzzViwYEGcccYZUVNTE4sXL47333+/qH23b98es2fPjurq6hg/fnw0NDRER0dHnxcNAJSPomLko48+iksvvTT2798f69evj4ceeihefPHFuPLKK+PIkSM97vvKK6/EvHnzYvLkyfHMM8/E8uXLY82aNbF8+fJ++QsAAANbUVdg/c53vhPvvvtu7Ny5M8aOHRsRERMmTIg5c+bEU089FQsXLjzuvoVCIerq6uKRRx6JysrKuOyyy+LTTz+Ne+65J+64444YM2ZM//xNAIABqagzI1u3bo1LL720M0QiImbPnh2TJ0+OzZs3H3e/tra22L59e1x77bVR+ZkvOVq0aFG0t7fHtm3bTmLpAEA5KCpGWlpaoq6urtv2urq6aGlpOe5+e/bsiba2tm77jh07NmpqanrcFwAYHIp6m6a1tTVGjBjRbfvIkSNjdw9fBtfa2hoRcdx9P/zww27bGxsbo7GxsfP2wYMHo6lEX9zT3t5ekmNPPnQoIv7w5U78QalmTXdmnQ9zzo9Z5yfVrIv+1t6Kiopu27Is63Gfo/f3Zt+lS5fG0qVLO2+PGzeuZN/W2FSqb4L841dd+5bJPynZrOnGrPNhzvkx6/ykmnVRb9PU1NR0nuX4rNbW1hg5cuRx9zt6X1/2BQAGh6JiZOrUqdHc3Nxte3Nzc0ydOvW4+9XW1kZVVVW3fQ8cOBCtra097gsADA5Fxcj8+fNjx44dceDAgc5tr776arz++uuxYMGC4+5XVVUV9fX18dhjj3V5W+bRRx+NoUOHxuWXX34SSwcAykFRMbJkyZIYPXp0fPWrX42tW7fGE088Eddcc03MmTMnrrrqqs7HFQqFqKioiL1793ZuW7lyZezatSsWL14cO3bsiHXr1sXKlSvj1ltvdY0RAKC4GDnzzDPj2WefjdGjR8c111wTN9xwQ1x00UWxZcuWLtcPOZY5c+bE1q1b47XXXosrrrgi7rvvvli2bFmsWrWqX/4CAMDAVvRv09TW1saWLVt6fEyhUIhCodBte319fdTX1/d6cQBA+fOtvQBAUkWfGaGX9uyJOMZVaweryYcOdV5/hdIy63yYc37MOkfr1yd5WmdGSmH37oja2tSrAIABwZmRUunhMvmD0a9cQTE3Zp0Pc86PWeco0WX3nRkBAJISIwBAUmIEAEhKjAAASYkRACApMQIAJCVGAICkxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJCUGAEAkhIjAEBSYgQASEqMAABJVWRZlqVeRE+GDRsWo0aNKsmxP/nkk/izP/uzkhybrsw6P2adD3POj1nnp5Szfu+99+Lw4cPHvO+Uj5FSGjduXOzbty/1MgYFs86PWefDnPNj1vlJNWtv0wAASYkRACCpQR0jS5cuTb2EQcOs82PW+TDn/Jh1flLNelB/ZgQASG9QnxkBANITIwBAUmUZI2+++WYsWLAgzjjjjKipqYnFixfH+++/X9S+27dvj9mzZ0d1dXWMHz8+GhoaoqOjo8QrHrj6MuuOjo5Ys2ZNXHLJJTF69Og466yz4qKLLoonn3wyn0UPQCfzM33U3r17Y/jw4VFRUREHDhwo0UoHvpOZ9eHDh+Ob3/xmTJw4MYYNGxbnnXde3HjjjSVe8cDV11n//ve/jzVr1sTUqVNj+PDhcf7558eNN94Y7777bg6rHnj27dsXt9xyS8ydOzeqq6ujoqKi6H1ze03MyszBgwez888/P5s1a1a2efPm7Ac/+EH253/+59mcOXOyjo6OHvd9+eWXs6FDh2Z/+Zd/mf3oRz/KHnzwwey0007Lli1bltPqB5a+zvrjjz/ORowYkd12223Z5s2bs23btmV//dd/nUVE9t3vfjfHv8HAcDI/05+1YMGC7JxzzskiItu/f38JVzxwncysjxw5ks2bNy8bM2ZM9u1vfzv78Y9/nD3++OPZ0qVLc1r9wHIys7777ruzz33uc9k3v/nN7Nlnn80efvjh7JxzzslmzZrVq38nBosdO3ZkY8aMyebPn59dfPHFWbEv/Xm+JpZdjKxevTobNmxY9pvf/KZz28svv5xFRPbv//7vPe47b968bMaMGV1+mFevXp0NHTo0O3DgQMnWPFD1dda///3vsw8//LDb9i9/+cvZlClTSrLWgexkfqaP2rRpU3b22Wdn//iP/yhGenAys/7nf/7nbMiQIVlTU1Opl1kWTmbWEyZMyBYvXtxl27/9279lEZG1tLSUZL0D2Wdf0+67776iYyTP18Sye5tm69atcemll8bYsWM7t82ePTsmT54cmzdvPu5+bW1tsX379rj22mujsvJPY1m0aFG0t7fHtm3bSrrugaivs/7c5z4XNTU13bbPmjUrfvOb35RkrQNZX+d81O9+97v427/927jvvvti5MiRpVzqgHcys37ooYfikksuiQsvvLDUyywLJzPr9vb2GDFiRJdtR29nfkG0m8++phUr79fEsouRlpaWqKur67a9rq4uWlpajrvfnj17oq2trdu+Y8eOjZqamh73Haz6OuvjeeGFF2Lq1Kn9sbSycrJzvvfee2P06NFxww03lGJ5ZaWvs25vb49XX301pkyZErfddlucddZZcdppp8UVV1wRb7zxRimXPGCdzM/1kiVL4l//9V9j27Zt8fHHH8euXbuioaEh6uvr/Tekn+T9mjikX492Cmhtbe1WzBERI0eOjN27d/e4X0Qcd98PP/ywv5ZYNvo662P5/ve/Hy+99FJs3Lixn1ZXPk5mzs3NzbF27dp4/vnne/WhtcGqr7P+4IMPor29Pf7lX/4lvvCFL8T69evjt7/9bdx1113xla98JZqbm6OqqqqEKx94TubneuXKldHR0RFXXHFF55mQyy67LJ544olSLHVQyvs1sezOjETEMf+je6JTd0fv78u+g1l/zOunP/1p/M3f/E381V/9VXzta1/rr6WVlb7O+aabborFixfHnDlzSrGsstSXWR/97YIjR47E5s2bY968eXH11VfHD37wg9izZ09s2LChJGsd6Pr6c/3ggw9GY2NjrFq1Kp577rn4/ve/H2+99VZ87WtfiyNHjpRiqYNO3q+JZXdmpKamprPoPqu1tbXH98uP3teXfQervs76s1577bW48sor45JLLomHHnqov5dYFvo658cffzx+/vOfx3e/+9343//934j4w+dHIiI++uijOOOMM+L0008vyZoHqr7OuqamJioqKuLCCy+Ms88+u3P7zJkz46yzzorm5uaSrHcg6+usP/jgg7jjjjvi7//+7+P222+PiIiLL744pk2bFrNnz46nn346/uIv/qJUyx408n5NLLszI1OnTj3mv/jNzc09vpdYW1sbVVVV3fY9cOBAtLa2eh/yGPo666PefvvtqK+vj0mTJsXGjRtjyJCya+N+0dc5t7S0xMcffxyTJk2KmpqaqKmpiZtvvjkiIi644AJnoY6hr7MePnx4TJgw4bj3Hzp0qD+WV1b6Ous9e/bE4cOHY+bMmV22z5gxIyoqKuK///u/+32tg1Her4llFyPz58+PHTt2dLmo06uvvhqvv/56LFiw4Lj7VVVVRX19fTz22GNdTkE9+uijMXTo0Lj88stLuu6BqK+zjoh49913o76+Ps4444zYsmVLDB8+vNTLHbD6OudvfOMbsWPHji5/li9fHhERTzzxRKxevbrkax9oTuZnev78+dHU1BTvvfde57af/exncfDgwZg1a1bJ1jxQ9XXWR6PvZz/7WZftr7zySmRZFp///OdLst7BJvfXxH79ReFTwNEL6cyePTvbsmVLtnHjxqy2trbbhXQaGhqyiMjeeuutzm0//elPs6FDh2bXX3999uyzz2bf/va3s+HDh7vo2XH0dda/+93vshkzZmSnnXZatmHDhuyll17q8ufQoUOJ/kanppP5mf6/vve977nOSA9OZtbvvPNOVlNTk82ZMyd76qmnsvXr12ef//zns2nTpvmZPoaTmfXVV1+dDR8+PPuHf/iH7Ec/+lH28MMPZ+PHj88mTpyY/fa3v03wtzn1bdiwIduwYUN2/fXXZxHReXv37t1ZlqV/TSy7GMmyLHvjjTeyK6+8Mjv99NOzs846K7v++uuz//mf/+nymOP9h/s//uM/slmzZmXDhg3LzjvvvGzFihVZe3t7jqsfWPoy67feeiuLiOP+6enFdLA6mZ/pzxIjJ3Yys/6v//qv7LLLLstOO+207Kyzzsquu+46s+5BX2f98ccfZ3fffXc2adKkrLq6Ojv//POzxYsXZ2+//XbOf4OB43j/vW1oaMiyLP1rYsUfFwkAkETZfWYEABhYxAgAkJQYAQCSEiMAQFJiBABISowAAEmJEQAgKTECACQlRgCApMQIAJDU/we1FZX5+ieqIQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 640x640 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "germ_cells = ['PGC', 'PGC_mitotic',  'GC_mitotic',  'oogonia_STRA8',  'oogonia_meiotic', 'pre_oocyte', 'oocyte', 'pre_spermatogonia']\n",
    "male_cl = ['Ti', 'pre_spermatogonia', 'FetalLeydig', 'Sertoli', 'late_sPAX8']\n",
    "female_cl = ['Oi',  'oogonia_STRA8',  'oogonia_meiotic', 'pre_oocyte', 'oocyte', 'preGC_I', 'preGC_IIa', 'preGC_IIb',  'granulosa', 'OSE']\n",
    "\n",
    "for sam in donors_match.keys():\n",
    "    print(sam)\n",
    "    andata_f = adata[[ind in donors_match[sam] for ind in adata.obs['donor']]]\n",
    "    freqs = andata_f.obs.groupby('labels').size()\n",
    "    # keep labels more than 5 cells\n",
    "#     if 'Hrv11' in sam:\n",
    "#         andata_f = andata_f[[i  in freqs[ freqs > 6].index for i in andata_f.obs.labels ]]\n",
    "#     else:\n",
    "    andata_f = andata_f[[i in freqs[ freqs > 3 ].index for i in andata_f.obs.labels ]] \n",
    "    andata_f = andata_f[[i in freqs[ freqs > 15 ].index or i in germ_cells for i in andata_f.obs.labels ]]        \n",
    "       \n",
    "    # F83 does not have any epithelial duct, from visual inspection of the H&E\n",
    "    if 'F83' in sam:\n",
    "        andata_f = andata_f[[ i not in 'Epithelial'  for i in andata_f.obs.labels ]] \n",
    "        andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]]  \n",
    "    if 'Hrv69' in sam:\n",
    "        andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]] \n",
    "    if 'Hrv58' in sam:\n",
    "        andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]] \n",
    "    if 'F91' in sam:\n",
    "        andata_f = andata_f[[ i not in ['granulosa']  for i in andata_f.obs.labels ]]\n",
    "        andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]]  \n",
    "    if 'Hrv11' in sam:\n",
    "        andata_f = andata_f[[ i not in ['granulosa']  for i in andata_f.obs.labels ]]\n",
    "        andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]]  \n",
    "    if 'F81' in sam:\n",
    "        andata_f = andata_f[[ i not in ['preGC_IIb', 'granulosa']  for i in andata_f.obs.labels ]]\n",
    "        andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]]     \n",
    "    if 'Hrv13' in sam:\n",
    "        andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]] \n",
    "    if 'F94' in sam:\n",
    "        andata_f = andata_f[[ i not in female_cl  for i in andata_f.obs.labels ]]  \n",
    "    if 'Hrv41' in sam:\n",
    "        andata_f = andata_f[[ i not in female_cl  for i in andata_f.obs.labels ]]  \n",
    "    if 'Hrv3' in sam:\n",
    "        andata_f = andata_f[[ i not in female_cl  for i in andata_f.obs.labels ]]   \n",
    "    if 'Hrv15' in sam:\n",
    "        andata_f = andata_f[[ i not in female_cl  for i in andata_f.obs.labels ]]   \n",
    "    if 'Hrv27' in sam:\n",
    "        andata_f = andata_f[[ i not in female_cl  for i in andata_f.obs.labels ]] \n",
    "# Reduce the number of genes by discarding lowly expressed genes\n",
    "# This is performed using 2 cut-off to remove as many lowly expressed genes as possible and yet avoid highly variable gene selection (HVG):\n",
    "# - Include all genes expressed by at least 1% of cells (cell_count_cutoff2)\n",
    "# - BECAUSE WE HAVE VERY RARE CELL TYPES (germ cells) Include genes expressed by at least 0.01% of cells (cell_count_cutoff) \n",
    "#   when they have high counts in non-zero cells (nonz_mean_cutoff)\n",
    "\n",
    "    # Remove cells and genes with 0 counts everywhere\n",
    "    sc.pp.filter_cells(andata_f, min_genes=2)\n",
    "    sc.pp.filter_genes(andata_f, min_cells=2)\n",
    "\n",
    "    # calculate the mean of each gene across non-zero cells\n",
    "    andata_f.var['n_cells'] = (andata_f.X.toarray() > 0).sum(0)\n",
    "    andata_f.var['nonz_mean'] = andata_f.X.toarray().sum(0) / andata_f.var['n_cells']\n",
    "    nonz_mean_cutoff = 0.05\n",
    "    cell_count_cutoff = np.log10(andata_f.shape[0] * 0.0001)\n",
    "    cell_count_cutoff2 = np.log10(andata_f.shape[0] * 0.01)\n",
    "    plt.vlines(nonz_mean_cutoff, cell_count_cutoff, cell_count_cutoff2, color='red');\n",
    "    plt.hlines(cell_count_cutoff, nonz_mean_cutoff, 1, color='red');\n",
    "    plt.hlines(cell_count_cutoff2, 0, nonz_mean_cutoff, color='red');\n",
    "\n",
    "    # select genes based on mean expression in non-zero cells\n",
    "    andata_ff = andata_f[:,(np.array(np.log10(andata_f.var['nonz_mean']) > nonz_mean_cutoff)\n",
    "                            | np.array(np.log10(andata_f.var['n_cells']) > cell_count_cutoff2))\n",
    "                             & np.array(np.log10(andata_f.var['n_cells']) > cell_count_cutoff)\n",
    "                             & np.array(~andata_f.var.index.isna())]\n",
    "\n",
    "    os.system('mkdir -p '+ path+'/cell2location/'+sam+'/')  \n",
    "    andata_ff.write(path + '/cell2location/' + sam + '/'+ sam + '_scRNAseq.h5ad')\n",
    "    \n",
    "    freqs = andata_ff.obs.groupby('labels').size()\n",
    "    print(freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Slice dataset for females"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "germ_cells = ['PGC', 'PGC_mitotic',  'GC', 'GC_mitotic',  'oogonia_STRA8',  'oogonia_meiotic',  'oocyte', 'pre_oocyte']\n",
    "sam = ['Hrv58', 'Hrv69', 'F83', 'F91', 'Hrv11']\n",
    "andata_f = adata[[ind in sam for ind in adata.obs['donor']]]\n",
    "freqs = andata_f.obs.groupby('labels').size()\n",
    "andata_f = andata_f[[i in freqs[ freqs > 50 ].index for i in andata_f.obs.labels ]] \n",
    "andata_f = andata_f[[i in freqs[ freqs > 40 ].index or i in germ_cells for i in andata_f.obs.labels ]]        \n",
    "\n",
    "\n",
    "# Remove male clusters\n",
    "andata_f = andata_f[[ i not in male_cl  for i in andata_f.obs.labels ]]  \n",
    "# Reduce the number of genes by discarding lowly expressed genes\n",
    "# This is performed using 2 cut-off to remove as many lowly expressed genes as possible and yet avoid highly variable gene selection (HVG):\n",
    "# - Include all genes expressed by at least 1% of cells (cell_count_cutoff2)\n",
    "# - BECAUSE WE HAVE VERY RARE CELL TYPES (germ cells) Include genes expressed by at least 0.01% of cells (cell_count_cutoff) \n",
    "#   when they have high counts in non-zero cells (nonz_mean_cutoff)\n",
    "\n",
    "# Remove cells and genes with 0 counts everywhere\n",
    "sc.pp.filter_cells(andata_f, min_genes=2)\n",
    "sc.pp.filter_genes(andata_f, min_cells=2)\n",
    "\n",
    "# calculate the mean of each gene across non-zero cells\n",
    "andata_f.var['n_cells'] = (andata_f.X.toarray() > 0).sum(0)\n",
    "andata_f.var['nonz_mean'] = andata_f.X.toarray().sum(0) / andata_f.var['n_cells']\n",
    "# plt.hist2d(np.log10(andata_f.var['nonz_mean']),\n",
    "#            np.log10(andata_f.var['n_cells']), bins=100,\n",
    "#            norm=mpl.colors.LogNorm(),\n",
    "#            range=[[0,0.5], [1,4.5]]);\n",
    "nonz_mean_cutoff = 0.05\n",
    "cell_count_cutoff = np.log10(andata_f.shape[0] * 0.0001)\n",
    "cell_count_cutoff2 = np.log10(andata_f.shape[0] * 0.01)\n",
    "plt.vlines(nonz_mean_cutoff, cell_count_cutoff, cell_count_cutoff2, color='red');\n",
    "plt.hlines(cell_count_cutoff, nonz_mean_cutoff, 1, color='red');\n",
    "plt.hlines(cell_count_cutoff2, 0, nonz_mean_cutoff, color='red');\n",
    "\n",
    "# select genes based on mean expression in non-zero cells\n",
    "andata_ff = andata_f[:,(np.array(np.log10(andata_f.var['nonz_mean']) > nonz_mean_cutoff)\n",
    "                        | np.array(np.log10(andata_f.var['n_cells']) > cell_count_cutoff2))\n",
    "                         & np.array(np.log10(andata_f.var['n_cells']) > cell_count_cutoff)\n",
    "                         & np.array(~andata_f.var.index.isna())]\n",
    "\n",
    "os.system('mkdir -p '+ path+'/cell2location/females/')  \n",
    "andata_ff.write(path + '/cell2location/females/females_scRNAseq.h5ad')\n",
    "\n",
    "freqs = andata_ff.obs.groupby('labels').size()\n",
    "print(freqs)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Slice dataset for males"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "germ_cells = ['sLGR5', 'PGC', 'PGC_mitotic',  'GC_mitotic',  'oogonia_STRA8',  'oogonia_meiotic',  'oocyte', 'gonocyte']\n",
    "sam = ['F94', 'Hrv41', 'Hrv3', 'Hrv27','Hrv15']\n",
    "andata_f = adata[[ind in sam for ind in adata.obs['donor']]]\n",
    "freqs = andata_f.obs.groupby('labels').size()\n",
    "andata_f = andata_f[[i in freqs[ freqs > 5 ].index for i in andata_f.obs.labels ]] \n",
    "andata_f = andata_f[[i in freqs[ freqs > 30 ].index or i in germ_cells for i in andata_f.obs.labels ]]        \n",
    "\n",
    "\n",
    "# Remove female clusters\n",
    "andata_f = andata_f[[ i not in female_cl  for i in andata_f.obs.labels ]] # Reduce the number of genes by discarding lowly expressed genes\n",
    "# This is performed using 2 cut-off to remove as many lowly expressed genes as possible and yet avoid highly variable gene selection (HVG):\n",
    "# - Include all genes expressed by at least 1% of cells (cell_count_cutoff2)\n",
    "# - BECAUSE WE HAVE VERY RARE CELL TYPES (germ cells) Include genes expressed by at least 0.01% of cells (cell_count_cutoff) \n",
    "#   when they have high counts in non-zero cells (nonz_mean_cutoff)\n",
    "\n",
    "# Remove cells and genes with 0 counts everywhere\n",
    "sc.pp.filter_cells(andata_f, min_genes=2)\n",
    "sc.pp.filter_genes(andata_f, min_cells=2)\n",
    "\n",
    "# calculate the mean of each gene across non-zero cells\n",
    "andata_f.var['n_cells'] = (andata_f.X.toarray() > 0).sum(0)\n",
    "andata_f.var['nonz_mean'] = andata_f.X.toarray().sum(0) / andata_f.var['n_cells']\n",
    "# plt.hist2d(np.log10(andata_f.var['nonz_mean']),\n",
    "#            np.log10(andata_f.var['n_cells']), bins=100,\n",
    "#            norm=mpl.colors.LogNorm(),\n",
    "#            range=[[0,0.5], [1,4.5]]);\n",
    "nonz_mean_cutoff = 0.05\n",
    "cell_count_cutoff = np.log10(andata_f.shape[0] * 0.0001)\n",
    "cell_count_cutoff2 = np.log10(andata_f.shape[0] * 0.01)\n",
    "plt.vlines(nonz_mean_cutoff, cell_count_cutoff, cell_count_cutoff2, color='red');\n",
    "plt.hlines(cell_count_cutoff, nonz_mean_cutoff, 1, color='red');\n",
    "plt.hlines(cell_count_cutoff2, 0, nonz_mean_cutoff, color='red');\n",
    "\n",
    "# select genes based on mean expression in non-zero cells\n",
    "andata_ff = andata_f[:,(np.array(np.log10(andata_f.var['nonz_mean']) > nonz_mean_cutoff)\n",
    "                        | np.array(np.log10(andata_f.var['n_cells']) > cell_count_cutoff2))\n",
    "                         & np.array(np.log10(andata_f.var['n_cells']) > cell_count_cutoff)\n",
    "                         & np.array(~andata_f.var.index.isna())]\n",
    "\n",
    "os.system('mkdir -p '+ path+'/cell2location/males/')  \n",
    "andata_ff.write(path + '/cell2location/males/males_scRNAseq.h5ad')\n",
    "\n",
    "freqs = andata_ff.obs.groupby('labels').size()\n",
    "print(freqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cellpymc",
   "language": "python",
   "name": "cellpymc"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
